### PR TITLE
Gradle Enterprise plugin degradation when using configuration cache

### PR DESCRIPTION
**Pull request by guylabs**
When using a version of the Gradle Enterprise plugin
< 3.12, and configuration cache is used, we'll disable
the Gradle Enterprise plugin and show a warning message.

⚠️ This can only be merged once `3.12` of the Gradle Enterprise plugin is released as the newly added test currently fails as the version is not yet published.

### CODE REVIEW

**Comment by guylabs** [comment]
@bot-gradle test

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginCheckInIntegrationTest.groovy [hunk 1]
```diff
@@ -86,4 +89,33 @@ class GradleEnterprisePluginCheckInIntegrationTest extends AbstractIntegrationSp
         then:
         plugin.serviceCreatedOnce(output)
     }
+
+    @IgnoreIf({ GradleContextualExecuter.configCache })
```
Comment (facewindu) [inline subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginCheckInIntegrationTest.groovy:93]: can you explain this ?
What does this executer do, and why do you need to remove the test from being executed with it ?
(forget my lack of knowledge here)

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/GradleEnterprisePluginServices.java [hunk 2]
```diff
@@ -61,7 +61,7 @@ public void registerGradleServices(ServiceRegistration registration) {
         registration.add(DefaultGradleEnterprisePluginBackgroundJobExecutors.class);
         registration.add(DefaultGradleEnterprisePluginBuildState.class);
         registration.add(DefaultGradleEnterprisePluginConfig.class);
-        registration.add(DefautGradleEnterprisePluginCheckInService.class);
+        registration.add(DefaultGradleEnterprisePluginCheckInService.class);
```
Comment (facewindu) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/GradleEnterprisePluginServices.java:64]: :)

**Review by facewindu — APPROVED** [review]

File: subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginCheckInIntegrationTest.groovy (hunk 1 above)
Comment (guylabs) [reply subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginCheckInIntegrationTest.groovy:93]: I took it from here:

https://github.com/gradle/gradle/blob/b0bb465ce06f49c2840016371893d006c09189ee/subprojects/enterprise/src/integTest/groovy/org/gradle/internal/enterprise/GradleEnterprisePluginConfigurationCachingIntegrationTest.groovy#L27-L29

**Comment by guylabs** [comment]
@bot-gradle test

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginCheckInService.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            }
+
+            @Override
+            public GradleEnterprisePluginServiceRef getPluginServiceRef() {
+                return pluginServiceRefSupplier.get();
+            }
+        };
+    }
+
+    private static boolean isUnsupported(String pluginVersion) {
```
Comment (ldaley) [inline subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginCheckInService.java:90]: This should be called something like `isUnsupportedWithConfigurationCaching()` or all of the checks should move in here.

**Review by ldaley — APPROVED** [review]

File: subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginCheckInService.java (hunk 3 above)
Comment (guylabs) [reply subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/DefaultGradleEnterprisePluginCheckInService.java]: Done in https://github.com/gradle/gradle/commit/0bcb7ddc86788fd6d0e45f6a99bc2c19161d893b

**Comment by guylabs** [comment]
@bot-gradle test and merge

**Review by marcphilipp — APPROVED** [review]

**Comment by ljacomet** [comment]
@bot-gradle test and merge

**Comment by guylabs** [comment]
> @bot-gradle test and merge

@ljacomet Thanks for pushing, but this will still fail as `3.12` of the Gradle Enterprise plugin is not yet released. Once released we can merge it. I'll take care of it.

**Comment by guylabs** [comment]
This is dependent on https://github.com/gradle/gradle/pull/23058

**Comment by guylabs** [comment]
@bot-gradle test and merge

**Comment by guylabs** [comment]
@bot-gradle test and merge

**Comment by guylabs** [comment]
@bot-gradle test and merge

**Pull request merged**